### ISSUE TITLE
Sessions working?

### ISSUE THREAD

**Issue by rafamds** [mattstauffer/Torch#52]
Hi, is the Sessions component still working, now with 5.4 version?

I tried to install it, but they never got stored on the right folder. I'm using all components on 5.4.

**Comment by dead23angel** [comment]
Hi, @rafamds!
My composer.json
```
{
    "require": {
        "slim/slim": "2.*",
        "zeuxisoo/slim-whoops": "~0.2.0",
        "filp/whoops": "1.1.2",
        "illuminate/session": "~5.4.0",
        "illuminate/support": "~5.4.0",
        "illuminate/container": "~5.4.0",
        "illuminate/config": "~5.4.0"
    }
}
```
I run command **composer install**
I run command **php -S localhost:8000**
I opened the **http://localhost:8000/set** in my browser and my session was correctly saved in the folder **sessions**

**Comment by rafamds** [comment]
I forgot to mention database driver, but thanks @dead23angel. I miss the bit where Matt says in the docs that only files driver is working

**Comment by dead23angel** [comment]
Hi, @rafamds!
Even using a database for storage, everything works correctly.
My composer.json
```
{
    "require": {
        "slim/slim": "2.*",
        "zeuxisoo/slim-whoops": "~0.2.0",
        "filp/whoops": "1.1.2",
        "illuminate/session": "~5.4.0",
        "illuminate/support": "~5.4.0",
        "illuminate/container": "~5.4.0",
        "illuminate/config": "~5.4.0",
	"illuminate/database": "~5.4.0",
        "illuminate/events": "~5.4.0"
    }
}
```

My index.php
```php
<?php

use Illuminate\Config\Repository as Config;
use Illuminate\Container\Container;
use Illuminate\Database\Capsule\Manager as Capsule;
use Illuminate\Events\Dispatcher;
use Illuminate\Filesystem\Filesystem;
use Illuminate\Session\SessionManager;
use Symfony\Component\HttpFoundation\Cookie;

require_once 'vendor/autoload.php';

/**
* Illuminate/session
*
* Illuminate Sessions outside of laravel;
*
* Requires: illuminate/support
*           illuminate/container
*           illuminate/session
*           illuminate/config
*
* @todo Drivers other than the file driver
*
* @source https://github.com/illuminate/session
* @contributor Matt Stauffer
* @contributor Sam Jordan
* @contributor Jordon Brill
*/

$app = new \Slim\Slim();
$app->add(new \Zeuxisoo\Whoops\Provider\Slim\WhoopsMiddleware);

// Init the container
$container = new Container;
$container->bind('app', $container);

$capsule = new Capsule;

$capsule->addConnection([
    'driver'    => 'mysql',
    'host'      => 'localhost',
    'database'  => 'db_name',
    'username'  => 'db_user',
    'password'  => 'db_password',
    'charset'   => 'utf8',
    'collation' => 'utf8_unicode_ci',
    'prefix'    => '',
]);

// Set the event dispatcher used by Eloquent models... (optional)
$capsule->setEventDispatcher(new Dispatcher(new Container));

// Set the cache manager instance used by connections... (optional)
// $capsule->setCacheManager(...);

// Make this Capsule instance available globally via static methods... (optional)
$capsule->setAsGlobal();

// Setup the Eloquent ORM... (optional; unless you've used setEventDispatcher())
$capsule->bootEloquent();

$container->bind('db', $capsule);

$container['config'] = new Config(require __DIR__ . '/config/app.php');

$container['files'] = new Filesystem;

// NOTE: some of the items for the cookie (lifetime, expire_on_close)
// are set in the config file and others are set below to show the
// flexibility when using config
$container['config']['session.table'] = 'sessions';
$container['config']['session.driver'] = 'database';

$container['config']['session.lottery'] = [2, 100]; // lottery--how often do they sweep storage location to clear old ones?
$container['config']['session.cookie'] = 'laravel_session';
$container['config']['session.path'] = '/';
$container['config']['session.domain'] = null;
// $container['config']['session.driver'] = 'file';
$container['config']['session.files'] = __DIR__ . '/sessions';

// Now we need to fire up the session manager
$sessionManager = new SessionManager($container);
$container['session.store'] = $sessionManager->driver();
$container['session'] = $sessionManager;

// In order to maintain the session between requests, we need to populate the
// session ID from the supplied cookie
$cookieName = $container['session']->getName();

if (isset($_COOKIE[$cookieName])) {
    if ($sessionId = $_COOKIE[$cookieName]) {
        $container['session']->setId($sessionId);
    }
}

// Boot the session
$container['session']->start();
// END BOOTSTRAP---------------------------------------------------------------

// View
$app->get('/', function () use ($container) {
    echo 'Current state of <code>$test</code>: ';

    if ($container['session']->has('test')) {
        echo 'Set<br> Value: ' . $container['session']->get('test');
    } else {
        echo 'Not set';
    }

    echo '<hr><a href="/set">Set session variable</a>';
});

// Set
$app->get('/set', function () use ($container) {
    $var = randomVar();

    if ($container['session']->has('test')) {
        echo '<p><code>$test</code> is set. Overriding it to now be <code>' . $var . '</code></p>';
    } else {
        echo '<p><code>$test</code> is not set. Setting to be <code>' . $var . '</code></p>';
    }

    $container['session']->put('test', $var);

    // Save the session
    $container['session']->save();

    // The session is saved; now, we'll store the session ID in a cookie to allow for
    // the session to remain on future requests
    $cookie = new Cookie(
        $container['session']->getName(),
        $container['session']->getId(),
        time() + ($container['config']['session.lifetime'] * 60),
        '/',
        null,
        false
    );

    setcookie(
        $cookie->getName(),
        $cookie->getValue(),
        $cookie->getExpiresTime(),
        $cookie->getPath(),
        $cookie->getDomain()
    );

    echo '<hr><a href="/">View current value of session variable</a>';
});

function randomVar() {
    $names = [
        "Afghanistan",
        "Åland Islands",
        "Albania",
        "Algeria",
        "American Samoa",
        "Andorra",
        "Angola",
        "Anguilla",
        "Antarctica",
        "Antigua and Barbuda",
        "Argentina",
        "Armenia",
        "Aruba",
        "Australia",
        "Austria",
        "Azerbaijan",
        "Bahamas",
        "Bahrain",
        "Bangladesh",
        "Barbados",
        "Belarus",
        "Belgium",
        "Belize",
        "Benin",
        "Bermuda",
        "Bhutan",
        "Bolivia (Plurinational State of)",
        "Bonaire, Sint Eustatius and Saba",
        "Bosnia and Herzegovina",
        "Botswana",
        "Bouvet Island",
        "Brazil",
        "British Indian Ocean Territory",
        "Brunei Darussalam",
        "Bulgaria",
        "Burkina Faso",
        "Burundi",
        "Cambodia",
        "Cameroon",
        "Canada",
        "Cabo Verde",
        "Cayman Islands",
        "Central African Republic",
        "Chad",
        "Chile",
        "China",
        "Christmas Island",
        "Cocos (Keeling) Islands",
        "Colombia",
        "Comoros",
        "Congo",
        "Congo (Democratic Republic of the)",
        "Cook Islands",
        "Costa Rica",
        "Côte d'Ivoire",
        "Croatia",
        "Cuba",
        "Curaçao",
        "Cyprus",
        "Czech Republic",
        "Denmark",
        "Djibouti",
        "Dominica",
        "Dominican Republic",
        "Ecuador",
        "Egypt",
        "El Salvador",
        "Equatorial Guinea",
        "Eritrea",
        "Estonia",
        "Ethiopia",
        "Falkland Islands (Malvinas)",
        "Faroe Islands",
        "Fiji",
        "Finland",
        "France",
        "French Guiana",
        "French Polynesia",
        "French Southern Territories",
        "Gabon",
        "Gambia",
        "Georgia",
        "Germany",
        "Ghana",
        "Gibraltar",
        "Greece",
        "Greenland",
        "Grenada",
        "Guadeloupe",
        "Guam",
        "Guatemala",
        "Guernsey",
        "Guinea",
        "Guinea-Bissau",
        "Guyana",
        "Haiti",
        "Heard Island and McDonald Islands",
        "Holy See",
        "Honduras",
        "Hong Kong",
        "Hungary",
        "Iceland",
        "India",
        "Indonesia",
        "Iran (Islamic Republic of)",
        "Iraq",
        "Ireland",
        "Isle of Man",
        "Israel",
        "Italy",
        "Jamaica",
        "Japan",
        "Jersey",
        "Jordan",
        "Kazakhstan",
        "Kenya",
        "Kiribati",
        "Korea (Democratic People's Republic of)",
        "Korea (Republic of)",
        "Kuwait",
        "Kyrgyzstan",
        "Lao People's Democratic Republic",
        "Latvia",
        "Lebanon",
        "Lesotho",
        "Liberia",
        "Libya",
        "Liechtenstein",
        "Lithuania",
        "Luxembourg",
        "Macao",
        "Macedonia (the former Yugoslav Republic of)",
        "Madagascar",
        "Malawi",
        "Malaysia",
        "Maldives",
        "Mali",
        "Malta",
        "Marshall Islands",
        "Martinique",
        "Mauritania",
        "Mauritius",
        "Mayotte",
        "Mexico",
        "Micronesia (Federated States of)",
        "Moldova (Republic of)",
        "Monaco",
        "Mongolia",
        "Montenegro",
        "Montserrat",
        "Morocco",
        "Mozambique",
        "Myanmar",
        "Namibia",
        "Nauru",
        "Nepal",
        "Netherlands",
        "New Caledonia",
        "New Zealand",
        "Nicaragua",
        "Niger",
        "Nigeria",
        "Niue",
        "Norfolk Island",
        "Northern Mariana Islands",
        "Norway",
        "Oman",
        "Pakistan",
        "Palau",
        "Palestine, State of",
        "Panama",
        "Papua New Guinea",
        "Paraguay",
        "Peru",
        "Philippines",
        "Pitcairn",
        "Poland",
        "Portugal",
        "Puerto Rico",
        "Qatar",
        "Réunion",
        "Romania",
        "Russian Federation",
        "Rwanda",
        "Saint Barthélemy",
        "Saint Helena, Ascension and Tristan da Cunha",
        "Saint Kitts and Nevis",
        "Saint Lucia",
        "Saint Martin (French part)",
        "Saint Pierre and Miquelon",
        "Saint Vincent and the Grenadines",
        "Samoa",
        "San Marino",
        "Sao Tome and Principe",
        "Saudi Arabia",
        "Senegal",
        "Serbia",
        "Seychelles",
        "Sierra Leone",
        "Singapore",
        "Sint Maarten (Dutch part)",
        "Slovakia",
        "Slovenia",
        "Solomon Islands",
        "Somalia",
        "South Africa",
        "South Georgia and the South Sandwich Islands",
        "South Sudan",
        "Spain",
        "Sri Lanka",
        "Sudan",
        "Suriname",
        "Svalbard and Jan Mayen",
        "Swaziland",
        "Sweden",
        "Switzerland",
        "Syrian Arab Republic",
        "Taiwan, Province of China",
        "Tajikistan",
        "Tanzania, United Republic of",
        "Thailand",
        "Timor-Leste",
        "Togo",
        "Tokelau",
        "Tonga",
        "Trinidad and Tobago",
        "Tunisia",
        "Turkey",
        "Turkmenistan",
        "Turks and Caicos Islands",
        "Tuvalu",
        "Uganda",
        "Ukraine",
        "United Arab Emirates",
        "United Kingdom of Great Britain and Northern Ireland",
        "United States of America",
        "United States Minor Outlying Islands",
        "Uruguay",
        "Uzbekistan",
        "Vanuatu",
        "Venezuela (Bolivarian Republic of)",
        "Viet Nam",
        "Virgin Islands (British)",
        "Virgin Islands (U.S.)",
        "Wallis and Futuna",
        "Western Sahara",
        "Yemen",
        "Zambia",
    ];

    return $names[array_rand($names)];
}

$app->run();
```

In the database, I created a **sessions** table using a template from the [documentation](https://laravel.com/docs/5.4/session#driver-prerequisites).

I run command composer install
I run command php -S localhost:8000
I opened the http://localhost:8000/set in my browser and my session was correctly saved in the sessions table

**Comment by huglester** [comment]
Hello,

when using file sessions. I don't get any errors. But session ID changes on every refresh - is it ok?

But when switching to database, like this:
```php
$container['config']['session.lottery'] = [2, 100]; // lottery
$container['config']['session.cookie'] = 'laravel_session';
$container['config']['session.path'] = '/';
$container['config']['session.domain'] = null;
//$container['config']['session.driver'] = 'file';
//$container['config']['session.files'] = __DIR__.'/sessions';
$container['config']['session.table'] = 'sessions';
$container['config']['session.driver'] = 'database';
```

I start getting similar errors:
```
Warning!
ErrorException [ Warning ]: Illegal offset type in isset or empty
VENDORPATH/illuminate/container/Container.php @ line 1021:
1020:	{
1021:	if (! isset($this->aliases[$abstract])) {
1022:	return $abstract;
Warning!
ErrorException [ Warning ]: Illegal offset type in isset or empty
VENDORPATH/illuminate/container/Container.php @ line 666:
665:	// type and then spin through them and check for contextual bindings on these.
666:	if (empty($this->abstractAliases[$abstract])) {
667:	return;
Warning!
ErrorException [ Warning ]: Illegal offset type in isset or empty
VENDORPATH/illuminate/container/Container.php @ line 588:
587:	// so the developer can keep using the same objects instance every time.
588:	if (isset($this->instances[$abstract]) && ! $needsContextualBuild) {
589:	return $this->instances[$abstract];
```

any idea what is happening?

thank you

**Comment by dead23angel** [comment]
Hi @huglester !
> when using file sessions. I don't get any errors. But session ID changes on every refresh - is it ok?
Did you try to specify these settings?
```
$container['config']['session.lifetime'] = 120;
$container['config']['session.expire_on_close'] = false;
```

I checked the session saving in the database, on components laravel 5.4 + php 7.0 + mysql 5.7

**Comment by huglester** [comment]
Hello @dead23angel thanks for helping!
I just took your composer.json and index.php

I also added config, as you asked to.. like so:

```php
$container['config']['session.lifetime'] = 120;
$container['config']['session.expire_on_close'] = false;
$container['config']['session.table'] = 'sessions';
$container['config']['session.driver'] = 'database';
```

I also created an empty config/app.php config returning empty array (return []);

and I still see these errors:
```php
Warning: Illegal offset type in isset or empty in /Users/huglester/tmp/sess/vendor/illuminate/container/Container.php on line 1021

Warning: Illegal offset type in isset or empty in /Users/huglester/tmp/sess/vendor/illuminate/container/Container.php on line 666

Warning: Illegal offset type in isset or empty in /Users/huglester/tmp/sess/vendor/illuminate/container/Container.php on line 588

Warning: Illegal offset type in isset or empty in /Users/huglester/tmp/sess/vendor/illuminate/container/Container.php on line 666

Warning: Illegal offset type in isset or empty in /Users/huglester/tmp/sess/vendor/illuminate/container/Container.php on line 644

Warning: Illegal offset type in /Users/huglester/tmp/sess/vendor/illuminate/container/Container.php on line 685

Warning: Illegal offset type in /Users/huglester/tmp/sess/vendor/illuminate/container/Container.php on line 685

Warning: Illegal offset type in isset or empty in /Users/huglester/tmp/sess/vendor/illuminate/container/Container.php on line 1021

Warning: Illegal offset type in isset or empty in /Users/huglester/tmp/sess/vendor/illuminate/container/Container.php on line 1042

Warning: Illegal offset type in isset or empty in /Users/huglester/tmp/sess/vendor/illuminate/container/Container.php on line 182

Warning: Illegal offset type in /Users/huglester/tmp/sess/vendor/illuminate/container/Container.php on line 183

Warning: Illegal offset type in /Users/huglester/tmp/sess/vendor/illuminate/container/Container.php on line 622
Current state of $test: Not set
```

Any ideas? are you sure I am not missing anything? how your config/app.php does look like?

Thanks

**Comment by huglester** [comment]
@dead23angel if you are able to, please attach composer.json with composer.lock files and other files. I am sure I am missing something! (btw I did try illuminate 5.4.* version - right?)

Thank you!

**Comment by dead23angel** [comment]
@huglester hi!
My test files, [download](https://yadi.sk/d/jec377_M3GvpY5)

**Comment by huglester** [comment]
Strange...

the same :( downloaded. composer install.. and I get this.
which exact php version you have?

php -S 127.0.0.1:800 

http://localhost:8000

and a results..:
![image](https://cloud.githubusercontent.com/assets/150731/25001974/85983bfc-2050-11e7-94e7-0cedb61f3929.png)

ANy more ideas?

**Comment by dead23angel** [comment]
I have an operating system Linux + PHP 7.0 + MYSQL 5.7

Minimal php 5.6.4

**Comment by huglester** [comment]
I have macos, with php 7.0.14, 5.7.11-4.
Strange it does not work at all...

**Comment by dead23angel** [comment]
@huglester 
Tried to run my example? Without transferring to your code

**Comment by huglester** [comment]
@dead23angel yes. I downloaded zip. composer install, run the PHP server.. and I get this error :(

Can you please share your php.ini ?
Maybe you have more ideas on what is going wrong?

**Comment by dead23angel** [comment]
@huglester [my php configs](https://yadi.sk/d/ZZDa2ZwU3H3YAH)

There are no ideas, all I can advise is to rent a cheap server or hosting and run the script there to make sure that the script is working.

**Comment by wordpresspluginpro** [comment]
@huglester @dead23angel 

I based my project off this repo, feel free to poke around all components are working with 5.4:
https://github.com/wordpresspluginpro/illumine-framework/blob/master/illumine/IlluminePlugin.php

**Comment by JeramyNS** [comment]
I had some problems trying to get sessions working on my wordpress plugin.

It was kinda re-creating the session on every refresh, so I figured it must be slightly incorrect config settings.
I tried adding in the extra config settings but that didn't actually seem to update the config set inside the session manager (accessed by running the method: "getSessionConfig") like it was completely ignoring some of the settings I added.

So I changed how I set the config settings to the following and it suddenly started working:
```
$app['config']->set('session', [
        'lottery'   => [2, 100], 
        'cookie'    => 'laravel_session',
        'path'      => '/',
        'domain'    => null,
        'encrypt'   => true,
        'driver'    => 'file',
        'expire_on_close' => false,
        'lifetime'  => 120,
        'connection' => null,
        'table'     => 'sessions',
        'secure'    => false,
        'files'     => $folder . 'storage/framework/sessions'
    ]);
```
Thought I'd post here incase there is something to it and it helps someone else out.

**Comment by mattstauffer** [comment]
If anyone still runs into any issues with this, please start a new issue. Thank you everyone who contributed here!!
